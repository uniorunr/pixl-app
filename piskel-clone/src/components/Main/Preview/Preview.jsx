import React, { Component } from 'react';
import PropTypes from 'prop-types';
import CanvasSize from './CanvasSize/CanvasSize';
import FpsControl from './FpsControl/FpsControl';
import './Preview.scss';

class Preview extends Component {
  constructor(props) {
    super(props);
    const { framesArray } = this.props;

    this.previewRef = React.createRef();

    this.state = {
      framesArray,
      fps: 12,
      currFrame: 0,
    };
  }

  componentDidMount = () => {
    this.animate();
  };

  animate = () => {
    const { fps } = this.state;
    setTimeout(this.changeFrame, 1000 / fps);
  };

  changeFrame = () => {
    const { framesArray, currFrame } = this.state;
    requestAnimationFrame(this.animate);
    const index = framesArray.indexOf(null);
    if (index !== -1) {
      framesArray.splice(index);
      this.setState({
        framesArray: [...framesArray],
        currFrame: 0,
      });
    }
    if (framesArray[currFrame]) {
      const canvas = this.previewRef.current;
      const ctx = canvas.getContext('2d');
      ctx.clearRect(0, 0, canvas.width, canvas.height);
      ctx.imageSmoothingEnabled = false;
      ctx.drawImage(framesArray[currFrame], 0, 0, canvas.width, canvas.height);
      this.setState({
        currFrame: currFrame + 1 <= framesArray.length - 1 ? currFrame + 1 : 0,
      });
    } else {
      this.setState({ currFrame: 0 });
    }
  };

  updateFps = (fps) => {
    this.setState({
      fps: +fps,
    });
  };

  handleFullScreen = () => {
    const previewCanvas = this.previewRef.current;
    previewCanvas.getContext('2d').imageSmoothingEnabled = false;
    const fullscreen = previewCanvas.requestFullscreen
      || previewCanvas.webkitRequestFullScreen
      || previewCanvas.mozRequestFullScreen;
    fullscreen.call(previewCanvas);
  };

  render() {
    const { handlePixelsPerCanvas } = this.props;
    const { fps } = this.state;

    return (
      <section className="preview-section">
        <div className="preview">
          <canvas
            id="preview-canvas"
            className="preview__canvas"
            width="128"
            height="128"
            ref={this.previewRef}
          />
          <button
            className="preview__fullscreen-button"
            type="button"
            onClick={this.handleFullScreen}
          >
            <i className="fas fa-expand" />
          </button>
          <FpsControl fps={fps} updateFps={this.updateFps} />
          <CanvasSize handlePixelsPerCanvas={handlePixelsPerCanvas} />
        </div>
      </section>
    );
  }
}

Preview.propTypes = {
  handlePixelsPerCanvas: PropTypes.func.isRequired,
  framesArray: PropTypes.instanceOf(Array).isRequired,
};

export default Preview;
