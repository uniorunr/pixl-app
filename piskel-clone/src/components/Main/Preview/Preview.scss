.preview-section {
  display: flex;
  justify-content: flex-start;
  flex-direction: column;
  align-items: center;
  margin: 0 20px;
  padding: 5px;
  font-weight: 700;
  font-size: 1.2rem;
  background-color: white;
  border: 2px solid black;
  cursor: context-menu;
  align-self: flex-start;

  .preview {
    position: relative;

    .preview__canvas {
      background: white;
      border: 2px solid black;
      image-rendering: pixelated;
    }

    .preview__fullscreen-button {
      position: absolute;
      top: 0;
      right: 0;
      border: 2px solid black;
      background: white;
      outline: none;
      cursor: pointer;
      font-size: 1.6rem;
      padding: 2px 5px;
    }
  }
}
