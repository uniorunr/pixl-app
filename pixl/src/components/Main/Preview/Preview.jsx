import React, { Component } from 'react';
import GIF from 'gif.js.optimized';
import download from 'downloadjs';
import PropTypes from 'prop-types';
import CanvasSize from './CanvasSize/CanvasSize';
import FpsControl from './FpsControl/FpsControl';
import DownloadModal from './DownloadModal/DownloadModal';
import './Preview.scss';

const canvasWithSize = (sourceCanvas, width, height) => {
  const tempCanvas = document.createElement('canvas');
  const tempContext = tempCanvas.getContext('2d');
  tempCanvas.width = width;
  tempCanvas.height = height;
  tempContext.imageSmoothingEnabled = false;
  tempContext.drawImage(sourceCanvas, 0, 0, width, height);
  return tempCanvas;
};

class Preview extends Component {
  constructor(props) {
    super(props);
    const { framesArray } = this.props;

    this.previewRef = React.createRef();
    this.timeOut = null;

    this.state = {
      framesArray,
      fps: 12,
      currFrame: 0,
      recording: false,
      gif: null,
      downloadModal: false,
      name: 'uniorunr',
      size: 500,
    };
  }

  componentDidMount = () => {
    this.animate();
  };

  componentWillUnmount() {
    cancelAnimationFrame(this.animate);
    clearTimeout(this.timeOut);
  }

  animate = () => {
    const { fps } = this.state;
    this.timeOut = setTimeout(this.changeFrame, 1000 / fps);
  };

  changeFrame = () => {
    const {
      framesArray, currFrame, fps, recording, gif, size,
    } = this.state;
    requestAnimationFrame(this.animate);
    const index = framesArray.indexOf(null);
    if (index !== -1) {
      framesArray.splice(index);
      this.setState({
        framesArray: [...framesArray],
        currFrame: 0,
      });
    }
    if (framesArray[currFrame] && this.previewRef.current) {
      const canvas = this.previewRef.current;
      const ctx = canvas.getContext('2d');
      ctx.clearRect(0, 0, canvas.width, canvas.height);
      ctx.imageSmoothingEnabled = false;
      ctx.drawImage(framesArray[currFrame], 0, 0, canvas.width, canvas.height);
      this.setState({
        currFrame: currFrame + 1 <= framesArray.length - 1 ? currFrame + 1 : 0,
      });
      if (recording && gif) {
        gif.addFrame(canvasWithSize(framesArray[currFrame], size, size), {
          copy: true,
          delay: 1000 / fps,
        });
        if (!framesArray[currFrame + 1]) {
          this.setState({
            recording: false,
          });
          gif.render();
        }
      }
    } else {
      this.setState({ currFrame: 0, recording: false });
    }
  };

  updateFps = (fps) => {
    this.setState({
      fps: +fps,
    });
  };

  download = (returnedName, returnedSize) => {
    const { name, size } = this.state;
    const finalName = returnedName || name;
    const finalDimensions = returnedSize || size;
    const gif = new GIF({
      workers: 2,
      quality: 10,
      width: finalDimensions,
      height: finalDimensions,
    });

    gif.on('finished', (blob) => {
      download(blob, `${finalName}.gif`, 'image/gif');
    });

    this.setState({
      recording: true,
      currFrame: 0,
      gif,
      size: finalDimensions,
      downloadModal: false,
    });
  };

  handleDownloadModal = () => {
    const { downloadModal } = this.state;
    this.setState({
      downloadModal: !downloadModal,
    });
  };

  handleFullScreen = () => {
    const previewCanvas = this.previewRef.current;
    previewCanvas.getContext('2d').imageSmoothingEnabled = false;
    const fullscreen = previewCanvas.requestFullscreen
      || previewCanvas.webkitRequestFullScreen
      || previewCanvas.mozRequestFullScreen;
    fullscreen.call(previewCanvas);
  };

  render() {
    const { handlePixelsPerCanvas } = this.props;
    const { fps, downloadModal } = this.state;

    return (
      <section className="preview-section">
        <div className="preview">
          <canvas
            id="preview-canvas"
            className="preview__canvas"
            width="128"
            height="128"
            ref={this.previewRef}
          />
          <button
            className="preview__download"
            type="button"
            onClick={this.handleDownloadModal}
          >
            <i className="fas fa-download" />
          </button>
          <button
            className="preview__fullscreen-button"
            type="button"
            onClick={this.handleFullScreen}
          >
            <i className="fas fa-expand" />
          </button>
          <FpsControl fps={fps} updateFps={this.updateFps} />
          <CanvasSize handlePixelsPerCanvas={handlePixelsPerCanvas} />
          {downloadModal ? (
            <DownloadModal
              closeDownloadModal={this.handleDownloadModal}
              download={this.download}
            />
          ) : null}
        </div>
      </section>
    );
  }
}

Preview.propTypes = {
  handlePixelsPerCanvas: PropTypes.func.isRequired,
  framesArray: PropTypes.instanceOf(Array).isRequired,
};

export default Preview;
