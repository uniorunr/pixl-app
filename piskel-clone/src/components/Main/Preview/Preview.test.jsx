import React from 'react';
import renderer from 'react-test-renderer';
import Preview from './Preview';

const mockHandlePixelsPerCanvas = () => {};

function createNodeMock(element) {
  if (element.type === 'input') {
    return {
      current: {
        value: 23,
      },
    };
  }
  return null;
}

it('renders correctly', () => {
  const options = { createNodeMock };
  const tree = renderer
    .create(
      <Preview
        framesArray={[]}
        handlePixelsPerCanvas={mockHandlePixelsPerCanvas}
      />,
      options,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
