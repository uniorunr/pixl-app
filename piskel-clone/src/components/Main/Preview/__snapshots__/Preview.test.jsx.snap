// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<section
  className="preview-section"
>
  <div
    className="preview"
  >
    <canvas
      className="preview__canvas"
      height="128"
      id="preview-canvas"
      width="128"
    />
    <button
      className="preview__fullscreen-button"
      onClick={[Function]}
      type="button"
    >
      <i
        className="fas fa-expand"
      />
    </button>
    <div
      className="fps-control"
    >
      <label
        htmlFor="fps-control__input"
      >
        <span
          className="fps-control__label"
        >
          FPS: 
          NaN
        </span>
        <input
          className="fps-control__input"
          defaultValue="12"
          id="fps-control__input"
          max="24"
          min="1"
          name="volume"
          onInput={[Function]}
          step="1"
          type="range"
        />
      </label>
    </div>
    <form
      className="canvas-size-controls"
      name="canvas-size-controls"
    >
      <label
        htmlFor="canvas-32"
      >
        32
        <input
          id="canvas-32"
          name="radio"
          onChange={[Function]}
          type="radio"
          value="32"
        />
      </label>
      <label
        htmlFor="canvas-64"
      >
        64
        <input
          defaultChecked={true}
          id="canvas-64"
          name="radio"
          onChange={[Function]}
          type="radio"
          value="64"
        />
      </label>
      <label
        htmlFor="canvas-128"
      >
        128
        <input
          id="canvas-128"
          name="radio"
          onChange={[Function]}
          type="radio"
          value="128"
        />
      </label>
    </form>
  </div>
</section>
`;
